lib/src/signing.rs

```diff
@@ -0,0 +1,217 @@
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(missing_docs)]
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::sync::RwLock;
+
+use thiserror::Error;
+
+use crate::backend::CommitId;
+use crate::settings::UserSettings;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum SigStatus {
+    /// Valid signature that matches the data.
+    Good,
+    /// Valid signature that could not be verified (e.g. due to an unknown key).
+    Unknown,
+    /// Valid signature that does not match the signed data.
+    Bad,
+    /// Invalid signature.
+    Invalid,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Verification {
+    /// The status of the signature.
+    pub status: SigStatus,
+    /// The key representation, if available. For GPG, this will be the key
+    /// fingerprint.
+    pub key: Option<String>,
+    /// A display string, if available. For GPG, this will be formatted primary
+    /// user ID.
+    pub display: Option<String>,
+}
+
+impl Verification {
+    pub fn unknown() -> Self {
+        Self {
+            status: SigStatus::Unknown,
+            key: None,
+            display: None,
+        }
+    }
+    pub fn invalid() -> Self {
+        Self {
+            status: SigStatus::Invalid,
+            key: None,
+            display: None,
+        }
+    }
+}
+
+/// The backend for signing and verifying cryptographic signatures.
+///
+/// This allows using different signers, such as GPG or SSH, or different
+/// versions of them.
+pub trait SigningBackend: Debug + Send + Sync {
+    /// The name of the backend, used for selecting the main backend in config
+    /// and in CLI.
+    fn name(&self) -> &str;
+
+    /// Check if the signature was created by this backend implementation.
+    ///
+    /// Should check the signature format, usually just looks at the prefix.
+    fn is_of_this_type(&self, signature: &[u8]) -> bool;
+
+    /// Create a signature for arbitrary data.
+    ///
+    /// The `key` parameter is what `jj sign` receives as key argument, or what
+    /// is configured in the `signing.key` config.
+    fn sign(&self, data: &[u8], key: Option<&str>) -> Result<Vec<u8>, SignError>;
+
+    /// Verify a signature. Should be reflexive with `sign`:
+    /// ```rust,ignore
+    /// verify(data, sign(data)?)?.status == SigStatus::Good
+    /// ```
+    fn verify(&self, data: &[u8], signature: &[u8]) -> Result<Verification, SignError>;
+}
+
+#[derive(Debug, Error)]
+pub enum SignError {
+    #[error("Would override a foreign commit signature")]
+    CannotRewrite,
```

Comment:
Is `CannotRewrite` a backend-specific error? (or remainder of old impl?)

Reply:
huh, yup, just a remainder, with new (simpler) design foreign signatures are just dropped (and will make cli print hints about those) so that rebases continue to never fail.

I've revised the error handling in the last version